core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,343 @@
+            fieldNode.put("type", schemaType);
+            fieldNode.put("description", field.descriptionInfo().docString());
+
+            // Fill required fields for the current object.
+            if (field.requirement() == FieldRequirement.REQUIRED) {
+                required.add(field.name());
+            }
+
+            if (field.typeSignature().type() == TypeSignatureType.ENUM) {
+                fieldNode.set("enum", getEnumType(field.typeSignature()));
+            }
+
+            // Set the current path to be "PREVIOUS_PATH/field.name".
+            final String currentPath = path + "/" + field.name();
+
+            // Only Struct types map to custom objects to we need reference to those structs.
+            // Having references to primitives do not make sense.
+            if (MEMORIZED_JSON_TYPES.contains(schemaType)) {
+                visited.put(fieldTypeSignature, currentPath);
+            }
+
+            if (field.typeSignature().type() == TypeSignatureType.MAP) {
+                generateMapFields(fieldNode, field, visited, currentPath + "/properties");
+            } else if (field.typeSignature().type() == TypeSignatureType.ITERABLE) {
+                generateArrayFields(fieldNode, field, visited, currentPath + "/properties");
+            } else {
+                generateStructFields(fieldNode, field, visited, currentPath + "/properties");
+            }
+        }
+
+        // Set current field inside the returned object.
+        root.set(field.name(), fieldNode);
+    }
+
+    /**
+     * Generate properties for the given fields and writes to the object node.
+     * @param fields list of fields that the child has.
+     * @param visited a map of visited fields, required for cycle detection.
+     * @param path current path as defined in JSON Schema spec, required for cyclic references.
+     * @param root object node that the results will be written to.
+     */
+    private void generateFields(List<FieldInfo> fields, Map<String, String> visited, String path,
+                                ObjectNode root) {
+        final ObjectNode objectNode = mapper.createObjectNode();
+        final ArrayNode required = mapper.createArrayNode();
+
+        for (FieldInfo field : fields) {
+            if (VALID_FIELD_LOCATIONS.contains(field.location())) {
+                generateField(field, visited, path, objectNode, required);
+            }
+        }
+
+        root.set("properties", objectNode);
+        root.set("required", required);
+    }
+
+    private void generateMapFields(ObjectNode fieldNode, FieldInfo field, Map<String, String> visited,
+                                   String path) {
+        final ObjectNode additionalProperties = mapper.createObjectNode();
+        final String nextPath = path + '/' + field.name();
+
+        // Keys are always converted to strings.
+        final TypeSignature valueType =
+                ((ContainerTypeSignature) field.typeSignature()).typeParameters().get(1);
```

Comment:
We can convert this to `MapTypeSignature`.
How about also returning right away if the key is not String?

Reply:
RPC JSON serialization converts every key type to string. I guess same thing happens to everything because every key in `JSON` should be string 😆 So, should we keep it that way for better compatability?

Reply:
We are going to use this for an annotated service and an annotated service theoretically can have a map whose key is not String. But probably there's no such case so let's keep it as is. 😄

Reply:
We can follow up with `Thrift`, `Annotated` and `GraphQL` (Maybe this is not necessary) tests.